src/ripple/app/misc/detail/WorkFile.h

```diff
@@ -0,0 +1,144 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2016 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_MISC_DETAIL_WORKFILE_H_INCLUDED
+#define RIPPLE_APP_MISC_DETAIL_WORKFILE_H_INCLUDED
+
+#include <ripple/app/misc/detail/Work.h>
+#include <boost/system/error_code.hpp>
+#include <cassert>
+
+namespace ripple {
+
+namespace detail {
+
+// Work with files
+class WorkFile: public Work
+    , public std::enable_shared_from_this<WorkFile>
+{
+protected:
+    using error_code = boost::system::error_code;
+    // Override the definition in Work.h
+    using response_type = std::string;
+
+public:
+    using callback_type =
+        std::function<void(error_code const&, response_type&&)>;
+public:
+    WorkFile(
+        std::string const& path,
+        boost::asio::io_service& ios, callback_type cb);
+    ~WorkFile();
+
+    void run() override;
+
+    void cancel() override;
+
+private:
+    std::string path_;
+    callback_type cb_;
+    boost::asio::io_service& ios_;
+    boost::asio::io_service::strand strand_;
+
+};
+
+//------------------------------------------------------------------------------
+
+WorkFile::WorkFile(
+    std::string const& path,
+    boost::asio::io_service& ios, callback_type cb)
+    : path_(path)
+    , cb_(std::move(cb))
+    , ios_(ios)
+    , strand_(ios)
+{
+}
+
+WorkFile::~WorkFile()
+{
+    if (cb_)
+        cb_ (make_error_code(boost::system::errc::interrupted),
+            {});
+}
+
+void
+WorkFile::run()
+{
+    if (! strand_.running_in_this_thread())
+        return ios_.post(strand_.wrap (std::bind(
+            &WorkFile::run, shared_from_this())));
+
+    error_code ec;
+    auto fileContents = [&]()
+        -> std::string
+    {
+        using namespace boost::filesystem;
+#if _MSC_VER    // MSVC
+        {
+            auto first = path_.find_first_not_of("\\/");
```

Comment:
Is the intent here to strip the leading `/` that is in a properly formatted URI? If so, I'd simply do:

    if (path_[0] == '/')
        path_ = path_.substr(1);

Regardless of whether you adopt this change or stick with what you have, what happens if the `path_` is empty after this manipulation?

Reply:
Yes, that's the intent. Windows file URLs look like `file:///C:/Lib/validators.txt`, so the path ends up with a leading `/`. If that's passed naively to a `std` file function, it doesn't like that, so I need to strip the leading `/`.

If the path ends up empty, then `canonical` converts that to the working directory. And since it's a directory, the `ifstream` can't load it, so it fails.

I don't know what I was thinking with the search, so I converted it to yours, _and_ changed the URL validation in `ValidatorSite::Site::Resource::Resource` to disallow a completely empty path in a file URL. This will disallow `file://`, but allow `file:///`... That doesn't seem to make much sense, does it?